pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,526 @@
+        // should be unreachable as diff must either be positive or negative
+        return Ok(().into());
+    }
+
+    pub fn request_delegate(
+        candidate: Candidate<T>,
+        delegator: Delegator<T>,
+        pool: TargetPool,
+        stake: T::Balance,
+    ) -> DispatchResultWithPostInfo {
+        ensure!(!stake.is_zero(), Error::<T>::StakeMustBeNonZero);
+
+        // Convert stake into joining shares quantity.
+        let shares = pools::Joining::<T>::stake_to_shares_or_init(&candidate, Stake(stake))?;
+
+        // If the amount was stake and is less than the value of 1 share it will round down to
+        // 0 share. We avoid doing any work for 0 shares.
+        ensure!(!shares.0.is_zero(), Error::<T>::StakeMustBeNonZero);
+
+        // We create the new joining shares. It returns the actual amount of stake those shares
+        // represents (due to rounding).
+        let stake = pools::Joining::<T>::add_shares(&candidate, &delegator, shares.clone())?;
+
+        // We hold the funds of the delegator and register its stake into the candidate stake.
+        T::Currency::hold(&T::CurrencyHoldReason::get(), &delegator, stake.0)?;
+        pools::Joining::<T>::increase_hold(&candidate, &delegator, &stake)?;
+        Candidates::<T>::add_total_stake(&candidate, &stake)?;
+
+        // We create/mutate a request for joining.
+        let now = frame_system::Pallet::<T>::block_number();
+        let operation_key = match pool {
+            TargetPool::AutoCompounding => PendingOperationKey::JoiningAutoCompounding {
+                candidate: candidate.clone(),
+                at_block: now,
+            },
+            TargetPool::ManualRewards => PendingOperationKey::JoiningManualRewards {
+                candidate: candidate.clone(),
+                at_block: now,
+            },
+        };
+
+        // We store/mutate the operation in storage.
+        let operation = PendingOperations::<T>::get(&delegator, &operation_key);
+        let operation = operation
+            .err_add(&shares.0)
+            .map_err(|_| Error::<T>::MathOverflow)?;
+        PendingOperations::<T>::set(&delegator, &operation_key, operation);
+
+        pools::check_candidate_consistency::<T>(&candidate)?;
+
+        Pallet::<T>::deposit_event(Event::<T>::RequestedDelegate {
+            candidate,
+            delegator,
+            towards: pool,
+            pending: stake.0,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn request_undelegate(
+        candidate: Candidate<T>,
+        delegator: Delegator<T>,
+        pool: TargetPool,
+        amount: SharesOrStake<T::Balance>,
+    ) -> DispatchResultWithPostInfo {
+        // Converts amount to shares of the correct pool
+        let shares = match (amount, pool) {
+            (SharesOrStake::Shares(s), _) => s,
+            (SharesOrStake::Stake(s), TargetPool::AutoCompounding) => {
+                pools::AutoCompounding::<T>::stake_to_shares(&candidate, Stake(s))?.0
```

Comment:
This rounds down to the number of shares, so it is possible for this method to return `Ok` when the `amount` is greater than the available amount:

67 shares at 1000000 per share = 67000000, but calling request_undelegate with 67999999 succeeds.

Obviously the returned amount is 67000000, but it may be confusing that the call succeeded.

Reply:
We should document that it is **up to** the provided amount, but it may use less.